update docs: Update GSoC application tips.

EmmalineLake opened the pull request:
Add missing link for GitHub.
Fix CZO and tell reader CZO stands for chat.zulip.org
Fix broken links to Matt Ringel's blog post.
Add link to Julia Evans blog post.
Add section heading for "Questions Are Important."
Rearrange some content to fit with new section heading.

EmmalineLake commented:
Question 1.

Line 154 passes ./tools/build-docs and ./tools/lint --fix, as well as all the tests, but it runs all the way across the screen instead of splitting in the way the other lines do.  Is this a problem, and if so, how do I fix it?

EmmalineLake commented:
Question 2.

When I first read this section a few weeks ago, I had to re-read it, because I didn't understand how it jumped from the paragraphs about a successful application to GSoC, ending line 132, and the section about asking questions starting on line 133.    There is a big logical jump between making a plausible and useful project proposal and the comments about asking questions in public.  

I'm wondering if the section about asking questions on CZO should be someplace else, or at the very least, have more of a lead-in.

timabbott commented:
> Line 154 passes ./tools/build-docs and ./tools/lint --fix, as well as all the tests, but it runs all the way across the screen instead of splitting in the way the other lines do. Is this a problem, and if so, how do I fix it?

The best solution for this is to use the Markdown reference syntax: https://www.markdownguide.org/basic-syntax/#reference-style-links

We use this in a bunch of places across our documentation (with somewhat different conventions).

timabbott commented:
> I'm wondering if the section about asking questions on CZO should be someplace else, or at the very least, have more of a lead-in.

It's probably worth adding a section heading.  That said, we could also think about moving it to /help/development-community or something like that.

timabbott commented:
Also, most editors have a line-wrapping feature that will cause them to reflow the content sensibly -- I'd try running that as part of this change.  Git's "color words" feature is nice for reviewing the changes visually when content got line-wrapped.

EmmalineLake commented:
>It's probably worth adding a section heading. That said, we could also think about moving it to /help/development-community or something like that.

I looked around the other documentation, and it seems like this is probably the best place for GSoC interns see this.  I added a heading and introduction, and then rearranged the items to flow together a little bit better.  I also added a link about asking questions, which I originally found on the Outreacy site.  It seemed like a good addition.

EmmalineLake commented:
>Also, most editors have a line-wrapping feature that will cause them to reflow the content sensibly -- I'd try running that as part of this change. Git's "color words" feature is nice for reviewing the changes visually when content got line-wrapped.

I think I have this turned on in VS code, but it didn't wrap this one.  I fixed it manually, according to the link above, and I'll do a little more looking about why it didn't work.

EmmalineLake commented:
I don't understand the error from the failed tests, and I'm not sure where to look it up.  I also asked at CZO #development.

```
Testing only internal links in documentation...
2021-11-16 19:33:27 [scrapy.downloadermiddlewares.retry] ERROR: Gave up retrying <GET file:///__w/zulip/zulip/docs/_build/html/#what-makes-a-great-zulip-contributor> (failed 3 times): [Errno 21] Is a directory: '/__w/zulip/zulip/docs/_build/html/'
2021-11-16 19:33:28 [scrapy.core.scraper] ERROR: Error downloading <GET file:///__w/zulip/zulip/docs/_build/html/#what-makes-a-great-zulip-contributor>
Traceback (most recent call last):
  File "/srv/zulip-venv-cache/47972165d4c4a088af2c1dd3c4345d80ad1ab6e3/zulip-py3-venv/lib/python3.6/site-packages/twisted/internet/defer.py", line 63, in run
    return f(*args, **kwargs)
  File "/srv/zulip-venv-cache/47972165d4c4a088af2c1dd3c4345d80ad1ab6e3/zulip-py3-venv/lib/python3.6/site-packages/scrapy/core/downloader/middleware.py", line 48, in process_request
    return (yield download_func(request=request, spider=spider))
  File "/srv/zulip-venv-cache/47972165d4c4a088af2c1dd3c4345d80ad1ab6e3/zulip-py3-venv/lib/python3.6/site-packages/twisted/internet/defer.py", line 191, in maybeDeferred
    result = f(*args, **kwargs)
  File "/srv/zulip-venv-cache/47972165d4c4a088af2c1dd3c4345d80ad1ab6e3/zulip-py3-venv/lib/python3.6/site-packages/scrapy/core/downloader/handlers/file.py", line 13, in download_request
    with open(filepath, 'rb') as fo:
IsADirectoryError: [Errno 21] Is a directory: '/__w/zulip/zulip/docs/_build/html/'
Failed!
Error: Process completed with exit code 1.
```

EmmalineLake commented:
I am working on running these tests in my local environment, but they stall out.  It might be my internet connection, so I'll try again when it's not so busy and report back tomorrow.

timabbott commented on docs/contributing/gsoc-ideas.md:
```diff
@@ -125,43 +123,57 @@ contributions throughout the application process, including after the
+  - [Try, then ask](https://www.mattringel.com/2013/09/30/you-must-try-and-then-you-must-ask/)
+  - [We aren’t just making code, we’re making history](https://www.harihareswara.net/sumana/2016/10/12/0)
+  - [How to ask Good Questions](https://jvns.ca/blog/good-questions/)
+- Understand [what makes a great Zulip contributor](..#what-makes-a-great-zulip-contributor)
```
I'm pretty sure your test failure is because the leading `..` here is invalid.

You can generally debug by trying to actually click links listed in the CI output.

EmmalineLake replied:
Okay.  I'm getting there.  Now I see how you found the error.  
I'll keep working on it.
Thanks!

EmmalineLake commented:
This should be ready.  The tests have passed once, except for the CodeQL, which is perpetually pending right now.  The last update only updated the commit message for e5746ba55a445cf54d7d5adb4317a69b22aa2f40, which should be the next-to last commit.

timabbott commented on docs/contributing/gsoc-ideas.md:
```diff
@@ -125,43 +123,57 @@ contributions throughout the application process, including after the
-questions](https://blogs.akamai.com/2013/10/you-must-try-and-then-you-must-ask.html),
-and developing and sharing project ideas and project proposals that
+contributions to Zulip projects, helping other applicants on [GitHub](https://github.com/zulip/zulip)
+and on [chat.zulip.org](https://chat.zulip.org) (CZO),
```
We prefer to always link to https://zulip.com/developer-community/ and avoid abbreviations for it, because the context on that page about how to behave is helpful (and a common problem is the flood of new GSoC students not having read the guidelines).  Also abbreviations are best to avoid since they require extra context to understand, so I try to never type them :)

timabbott commented:
Merged, after the edit noted above and a bit more copy-editing (E.g. I decided "Zulip public stream" was less cumbersome than "public stream on chat.zulip.org").  

Huge thanks for doing this -- I think it'll make a big difference.

timabbott commented:
One more note: I fixed a few places where you had trailing whitespace at the end of lines while reviewing this; generally you should configure your editor to avoid generating those.

(We have linter checks for this in code, but apparently not in the documentation).
